lib/mempool: Fix spurious -ENOMEM due to agressive latency control

andyross opened the pull request:
The mempool operations need to be atomic, but because of latency
concerns (the allocator is intended for use in an ISR) the locking was
designed to be as minimal as possible.  And it... mostly got it right.
All the list handling was correctly synchronized.  The merging of four
child blocks into a parent block was atomic.  The splitting of a block
into four children was atomic.

BUT: there was a moment between the allocation of a large block and
the re-addition of its three unused children where the lock was being
released.  This meant that another context (e.g. an ISR that just
fired, interrupting the existing call to k_mem_pool_alloc()) would see
some memory "missing" that wasn't actually allocated.  And if this
happens to have been the top level block, it's entirely possible that
the whole heap looks empty, even though the other allocator might have
been doing only the smallest allocation!

Fix that by making the "remove a block then add back the three
children we don't use" into an atomic step.  We can still relax the
lock between levels as we split the subblocks further.

(Finally, note that this trick allows a somewhat cleaner API as we can
do our "retry due to race" step internally by walking back up the
block size list instead of forcing our caller to do it via that weird
-EAGAIN return value.)

Fixes #11022

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
lib/mempool: Fix spurious -ENOMEM due to agressive latency control

The mempool operations need to be atomic, but because of latency
concerns (the allocator is intended for use in an ISR) the locking was
designed to be as minimal as possible.  And it... mostly got it right.
All the list handling was correctly synchronized.  The merging of four
child blocks into a parent block was atomic.  The splitting of a block
into four children was atomic.

BUT: there was a moment between the allocation of a large block and
the re-addition of its three unused children where the lock was being
released.  This meant that another context (e.g. an ISR that just
fired, interrupting the existing call to k_mem_pool_alloc()) would see
some memory "missing" that wasn't actually allocated.  And if this
happens to have been the top level block, it's entirely possible that
the whole heap looks empty, even though the other allocator might have
been doing only the smallest allocation!

Fix that by making the "remove a block then add back the three
children we don't use" into an atomic step.  We can still relax the
lock between levels as we split the subblocks further.

(Finally, note that this trick allows a somewhat cleaner API as we can
do our "retry due to race" step internally by walking back up the
block size list instead of forcing our caller to do it via that weird
-EAGAIN return value.)

Fixes #11022

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

ceolin commented on lib/mempool/mempool.c:
```diff
@@ -255,26 +249,36 @@ int _sys_mem_pool_block_alloc(struct sys_mem_pool_base *p, size_t size,
 	}
+	pool_irq_unlock(p, key);
 
 	*level_p = alloc_l;
```
level_p is unsigned but alloc_l is signed and if there is no available block its value will be -1.

ceolin replied:
ok, just saw that this is not possible, though, you're still assigning an unsigned to signed. Either you cast it or change alloc_l to be signed (requiring some logic change).

ceolin reviewed (approved):
The patch itself doesn't introduced the problem I've mentioned. So +1

pdunaj reviewed (changes requested):
It seems race with two allocations will be fixed. There is still a case with `block_free` vs `_sys_mem_pool_block_alloc` race.
If interrupt will come while thread is in line 180 whole memory can disappear and allocation will fail.

carlescufi commented:
@andyross can you take a look at [this comment](https://github.com/zephyrproject-rtos/zephyr/pull/11220#pullrequestreview-173375489)?

andyross commented:
I see it and agree.   The "remove three sibling blocks and add back the combined one" needs to be atomic and it isn't currently (there is a gap between the removal and addition, so an unlocky allocator at that moment will see missing memory).  Needs a little rework to how the merge loop is written, because it defeats the existing tail recursion trick I was so proud of.

carlescufi commented:
@andyross thanks for the fix so far. Any idea when you will be able to take a look at the additional failure? Folks at Nordic are hitting this bug. Thanks!

Commit:
lib/mempool: One more spurious -ENOMEM condition

Similar to the last patch, there was a spot in block recombination
where the lock would be released while the combined block was being
held allocated.  That means that when recombining a single top-level
block, it was possible for the entire heap to look allocated.

Make the combination and re-addition of the larger block atomic.
Requires a little surgery to the structure of the code, so this is a
little more involved than the earlier fix.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Rebased.  Added an extra patch for the extra case @pdunaj caught.

pdunaj commented:
@andyross , thanks! This look ok now and should work. Let me quickly test it...

pdunaj commented:
I have checked the patch and it works fine :)
Thanks again @andyross .

pdunaj reviewed (approved):
I suggest squashing two commits into one.

carlescufi commented:
@pdunaj I actually think the 2 commits make sense, since they fix 2 different issues. I will merge as-is.
